File: Runtime/OutputStratumTransition.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

''' <summary>
''' The Output Stratum Transition class
''' </summary>
''' <remarks></remarks>
Friend Class OutputStratumTransition

    Private m_StratumId As Integer
    Private m_SecondaryStratumId As Nullable(Of Integer)
    Private m_Iteration As Integer
    Private m_Timestep As Integer
    Private m_TransitionGroupId As Integer
    Private m_AgeMin As Nullable(Of Integer)
    Private m_AgeMax As Nullable(Of Integer)
    Private m_AgeKey As Integer
    Private m_Amount As Double

    Public Sub New(
        ByVal stratumId As Integer,
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal iteration As Integer,
        ByVal timestep As Integer,
        ByVal transitionGroupId As Integer,
        ByVal ageMin As Nullable(Of Integer),
        ByVal ageMax As Nullable(Of Integer),
        ByVal ageKey As Integer,
        ByVal amount As Double)

        Me.m_StratumId = stratumId
        Me.m_SecondaryStratumId = secondaryStratumId
        Me.m_Iteration = iteration
        Me.m_Timestep = timestep
        Me.m_TransitionGroupId = transitionGroupId
        Me.m_AgeMin = ageMin
        Me.m_AgeMax = ageMax
        Me.m_AgeKey = ageKey
        Me.m_Amount = amount

    End Sub

    ''' <summary>
    ''' The output Stratum ID
    ''' </summary>
    Public ReadOnly Property StratumId As Integer
        Get
            Return Me.m_StratumId
        End Get
    End Property

    ''' <summary>
    ''' Gets the secondary stratum Id
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property SecondaryStratumId As Nullable(Of Integer)
        Get
            Return Me.m_SecondaryStratumId
        End Get
    End Property

    ''' <summary>
    ''' The output iteration
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property Iteration As Integer
        Get
            Return Me.m_Iteration
        End Get
    End Property

    ''' <summary>
    ''' The output timestep
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property Timestep As Integer
        Get
            Return Me.m_Timestep
        End Get
    End Property

    ''' <summary>
    ''' The output Transition Group ID
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property TransitionGroupId As Integer
        Get
            Return Me.m_TransitionGroupId
        End Get
    End Property

    ''' <summary>
    ''' Gets the minimum age
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property AgeMin As Nullable(Of Integer)
        Get
            Return Me.m_AgeMin
        End Get
    End Property

    ''' <summary>
    ''' Gets the maximum age
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property AgeMax As Nullable(Of Integer)
        Get
            Return Me.m_AgeMax
        End Get
    End Property

    ''' <summary>
    ''' Gets the age key
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public ReadOnly Property AgeKey As Integer
        Get
            Return Me.m_AgeKey
        End Get
    End Property

    ''' <summary>
    ''' The output amount
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Property Amount As Double
        Get
            Return Me.m_Amount
        End Get
        Set(ByVal value As Double)
            Me.m_Amount = value
        End Set
    End Property

End Class

File: src/Runtime/OutputStratumTransition.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2021 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System.Diagnostics;

namespace SyncroSim.STSim
{
    internal class OutputStratumTransition
    {
        private int m_StratumId;
        private int? m_SecondaryStratumId;
        private int? m_TertiaryStratumId;
        private int m_Iteration;
        private int m_Timestep;
        private int m_TransitionGroupId;
        private int? m_AgeMin;
        private int? m_AgeMax;
        private int m_AgeKey;
        private int? m_EventId;
        private int m_EventIdKey;
        private double m_Amount;

        public OutputStratumTransition(
            int stratumId, int? secondaryStratumId, int? tertiaryStratumId, int iteration, int timestep, 
            int transitionGroupId, int? ageMin, int? ageMax, int ageKey, int? eventId, int eventIdKey, double amount)
        {
            this.m_StratumId = stratumId;
            this.m_SecondaryStratumId = secondaryStratumId;
            this.m_TertiaryStratumId = tertiaryStratumId;
            this.m_Iteration = iteration;
            this.m_Timestep = timestep;
            this.m_TransitionGroupId = transitionGroupId;
            this.m_AgeMin = ageMin;
            this.m_AgeMax = ageMax;
            this.m_AgeKey = ageKey;
            this.m_EventId = eventId;
            this.m_EventIdKey = eventIdKey;
            this.m_Amount = amount;

#if DEBUG
            if (EventId.HasValue)
            {
                Debug.Assert(EventId.Value > 0);
            }
            else
            {
                Debug.Assert(EventIdKey == 0);
            }
#endif
        }

        /// <summary>
        /// The output Stratum ID
        /// </summary>
        public int StratumId
        {
            get
            {
                return this.m_StratumId;
            }
        }

        /// <summary>
        /// Gets the secondary stratum Id
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int? SecondaryStratumId
        {
            get
            {
                return this.m_SecondaryStratumId;
            }
        }

        /// <summary>
        /// Gets the tertiary stratum Id
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int? TertiaryStratumId
        {
            get
            {
                return this.m_TertiaryStratumId;
            }
        }

        /// <summary>
        /// The output iteration
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int Iteration
        {
            get
            {
                return this.m_Iteration;
            }
        }

        /// <summary>
        /// The output timestep
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int Timestep
        {
            get
            {
                return this.m_Timestep;
            }
        }

        /// <summary>
        /// The output Transition Group ID
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int TransitionGroupId
        {
            get
            {
                return this.m_TransitionGroupId;
            }
        }

        /// <summary>
        /// Gets the minimum age
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int? AgeMin
        {
            get
            {
                return this.m_AgeMin;
            }
        }

        /// <summary>
        /// Gets the maximum age
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int? AgeMax
        {
            get
            {
                return this.m_AgeMax;
            }
        }

        /// <summary>
        /// Gets the age key
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int AgeKey
        {
            get
            {
                return this.m_AgeKey;
            }
        }

        /// <summary>
        /// Gets the event Id
        /// </summary>
        public int? EventId
        {
            get
            {
                return this.m_EventId;
            }
        }

        /// <summary>
        /// Gets the event Id key
        /// </summary>
        public int EventIdKey
        {
            get
            {
                return this.m_EventIdKey;
            }
        }

        /// <summary>
        /// The output amount
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public double Amount
        {
            get
            {
                return this.m_Amount;
            }
            set
            {
                this.m_Amount = value;
            }
        }
    }
}
